File: ssm-sdk-client/src/test/java/ssm/client/SsmClientItTest.java
package ssm.client;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

import org.assertj.core.util.Lists;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;

import com.google.common.collect.ImmutableMap;

import ssm.client.sign.model.Signer;
import ssm.dsl.*;
import ssm.dsl.blockchain.Block;
import ssm.dsl.blockchain.Transaction;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class SsmClientItTest {

    private static String uuid = UUID.randomUUID().toString();
    private static final String NETWORK = "bclan-it/";
    public static final String ADMIN_NAME = "ssm-admin";
    public static final String USER1_NAME = "bob" + "-" + uuid;
    public static final String USER2_NAME = "sam" + "-" + uuid;
    public static final String USER1_FILENAME = NETWORK + "bob";
    public static final String USER2_FILENAME = NETWORK + "sam";

//    private static final String NETWORK = "bc1/";
//    private static final String ADMIN_NAME = "adrien";
//    private static final String USER1_NAME = "chuck";
//    private static final String USER2_NAME = "sarah";
//    private static final String USER1_FILENAME = NETWORK+"chuck";
//    private static final String USER2_FILENAME = NETWORK+"sarah";

    private static Signer signerAdmin;
    private static Signer signerUser1;
    private static Signer signerUser2;

    private static SsmAgentBase agentAdmin;
    private static SsmAgentBase agentUser1;
    private static SsmAgentBase agentUser2;

    private static SsmClient client;
    private static String ssmName;
    private static String sessionName;
    private static SsmSessionBase session;

    @BeforeAll
    public static void init() throws Exception {
        signerAdmin = Signer.Companion.loadFromFile(ADMIN_NAME, NETWORK + ADMIN_NAME);
        signerUser1 = Signer.Companion.loadFromFile(USER1_NAME, USER1_FILENAME);
        signerUser2 = Signer.Companion.loadFromFile(USER2_NAME, USER2_FILENAME);

        agentAdmin = AgentUtils.loadFromFile(ADMIN_NAME, NETWORK + ADMIN_NAME);
        agentUser1 = AgentUtils.loadFromFile(signerUser1.getName(), USER1_FILENAME);
        agentUser2 = AgentUtils.loadFromFile(signerUser2.getName(), USER2_FILENAME);

        client = SsmClientTestBuilder.build();

        ssmName = "CarDealership-" + uuid;
        Map<String, String> roles = ImmutableMap.of(signerUser1.getName(), "Buyer", signerUser2.getName(), "Seller");
        sessionName = "deal20181201-" + uuid;
        session = new SsmSessionBase(ssmName, sessionName, roles,"Used car for 100 dollars.", new HashMap());
    }

    @AfterEach
    public void waitBetweenTest() throws InterruptedException {
        //Node rest api return http response before the transaction had been mined
        Thread.sleep(2000);
    }

    @Test
    @Order(5)
    public void getListAdmin() throws Exception {
        CompletableFuture<List<String>> agentRet = client.listAdmins();
        List<String> agentFormClient = agentRet.get();
        assertThat(agentFormClient).contains(ADMIN_NAME);
    }


    @Test
    @Order(10)
    public void getAdminUser() throws Exception {
        CompletableFuture<Optional<SsmAgentBase>> agentRet = client.getAdmin(ADMIN_NAME);
        Optional<SsmAgentBase> agentFormClient = agentRet.get();
        assertThat(agentFormClient.get()).isEqualTo(agentAdmin);
    }

    @Test
    @Order(20)
    public void registerUser1() throws Exception {
        CompletableFuture<InvokeReturn> transactionEvent = client.registerUser(signerAdmin, agentUser1);
        InvokeReturn trans = transactionEvent.get();

        assertThatTransactionExists(trans);

    }

    @Test
    @Order(30)
    public void getAgentUser1() throws Exception {
        CompletableFuture<Optional<SsmAgentBase>> agentRet = client.getAgent(agentUser1.getName());
        Optional<SsmAgentBase> agentFormClient = agentRet.get();
        assertThat(agentFormClient.get()).isEqualTo(agentUser1);
    }

    @Test
    @Order(40)
    public void registerUser2() throws Exception {
        CompletableFuture<InvokeReturn> transactionEvent = client.registerUser(signerAdmin, agentUser2);
        InvokeReturn trans = transactionEvent.get();
        assertThatTransactionExists(trans);
    }

    @Test
    @Order(50)
    public void getAgentUser2() throws Exception {
        CompletableFuture<Optional<SsmAgentBase>> agentRet = client.getAgent(agentUser2.getName());
        Optional<SsmAgentBase> agentFormClient = agentRet.get();
        assertThat(agentFormClient.get()).isEqualTo(agentUser2);
    }

    @Test
    @Order(55)
    public void listAgent() throws Exception {
        CompletableFuture<List<String>> agentRet = client.listAgent();
        List<String> agentFormClient = agentRet.get();
        assertThat(agentFormClient).contains(agentUser1.getName(), agentUser2.getName());
    }

    @Test
    @Order(60)
    public void createSsm() throws Exception {
        SsmTransitionBase sell = new SsmTransitionBase(0, 1, "Seller", "Sell");
        SsmTransitionBase buy = new SsmTransitionBase(1, 2, "Buyer", "Buy");
        SsmBase ssm = new SsmBase(ssmName, Lists.newArrayList(sell, buy).toArray(new SsmTransitionBase[2]));
        CompletableFuture<InvokeReturn> transactionEvent = client.create(signerAdmin, ssm);
        InvokeReturn trans = transactionEvent.get();
        assertThatTransactionExists(trans);
    }

    @Test
    @Order(70)
    public void getSsm() throws Exception {
        CompletableFuture<Optional<SsmBase>> ssmReq = client.getSsm(ssmName);
        Optional<SsmBase> ssm = ssmReq.get();
        assertThat(ssm).isPresent();
        assertThat(ssm.get().getName()).isEqualTo(ssmName);
    }

    @Test
    @Order(80)
    public void start() throws Exception {
        Map<String, String> roles = ImmutableMap.of(agentUser1.getName(), "Buyer", agentUser2.getName(), "Seller");
        SsmSessionBase session = new SsmSessionBase(ssmName, sessionName, roles, "Used car for 100 dollars.", new HashMap());

        CompletableFuture<InvokeReturn> transactionEvent = client.start(signerAdmin, session);
        InvokeReturn trans = transactionEvent.get();
        assertThatTransactionExists(trans);
    }

    @Test
    @Order(90)
    public void getSession() throws Exception {
        CompletableFuture<Optional<SsmSessionStateBase>> ses = client.getSession(sessionName);
        Optional<SsmSessionStateBase> sesReq = ses.get();

        assertThat(sesReq.get().getCurrent()).isEqualTo(0);
        assertThat(sesReq.get().getIteration()).isEqualTo(0);
        assertThat(sesReq.get().getOrigin()).isNull();

        assertThat(sesReq.get().getSsm()).isEqualTo(ssmName);
        assertThat(sesReq.get().getRoles()).isEqualTo(session.getRoles());
        assertThat(sesReq.get().getSession()).isEqualTo(session.getSession());
        assertThat(sesReq.get().getPublic()).isEqualTo(session.getPublic());

    }

    private static Map<String, String> privateMessage;

    @Test
    @Order(100)
    public void performSell() throws Exception {
        SsmContextBase context = new SsmContextBase(sessionName, "100 dollars 1978 Camaro", 0, new HashMap());
        context = PrivateMessageUtils.addPrivateMessage(context, "Message to signer1", agentUser1);
        privateMessage = context.getPrivate();
        CompletableFuture<InvokeReturn> transactionEvent = client.perform(signerUser2, "Sell", context);
        InvokeReturn trans = transactionEvent.get();
        assertThatTransactionExists(trans);
    }


    @Test
    @Order(110)
    public void getSessionAfterSell() throws Exception {
        SsmTransitionBase sell = new SsmTransitionBase(0, 1, "Seller", "Sell");
        CompletableFuture<Optional<SsmSessionStateBase>> sesReq = client.getSession(sessionName);
        Optional<SsmSessionStateBase> state = sesReq.get();
        SsmSessionStateBase stateExpected = new SsmSessionStateBase(ssmName, sessionName, session.getRoles(), "100 dollars 1978 Camaro", privateMessage, sell, 1, 1);
        assertThat(state.get()).isEqualTo(stateExpected);

        String expectedMessage = PrivateMessageUtils.getPrivateMessage(stateExpected, signerUser1);
    }

    @Test
    @Order(110)
    public void getSessionAfterSellShouldReturnEncryptMessage() throws Exception {
        SsmTransitionBase sell = new SsmTransitionBase(0, 1, "Seller", "Sell");
        CompletableFuture<Optional<SsmSessionStateBase>> sesReq = client.getSession(sessionName);
        Optional<SsmSessionStateBase> state = sesReq.get();
        String expectedMessage = PrivateMessageUtils.getPrivateMessage(state.get(), signerUser1);
        assertThat(expectedMessage).isEqualTo("Message to signer1");

    }

    @Test
    @Order(120)
    public void performBuy() throws Exception {
        SsmContextBase context = new SsmContextBase(sessionName, "Deal !", 1, new HashMap());
        CompletableFuture<InvokeReturn> transactionEvent = client.perform(signerUser1, "Buy", context);
        InvokeReturn trans = transactionEvent.get();
        assertThatTransactionExists(trans);
    }

    void assertThatTransactionExists(InvokeReturn trans) throws ExecutionException, InterruptedException {
        assertThat(trans).isNotNull();
        assertThat(trans.getStatus()).isEqualTo("SUCCESS");
        Transaction transaction = client.getTransaction(trans.getTransactionId()).get().orElse(null);
        assertThat(transaction).isNotNull();
        assertThat(transaction.getBlockId()).isNotNull();
        Block block = client.getBlock(transaction.getBlockId()).get().orElse(null);
        assertThat(block).isNotNull();
    }

    @Test
    @Order(130)
    public void getSessionAfterBuy() throws Exception {
        SsmTransitionBase buy = new SsmTransitionBase(1, 2, "Buyer", "Buy");
        CompletableFuture<Optional<SsmSessionStateBase>> sesReq = client.getSession(sessionName);
        Optional<SsmSessionStateBase> state = sesReq.get();
        SsmSessionStateBase stateExcpected = new SsmSessionStateBase(ssmName, sessionName, session.getRoles(),"Deal !", new HashMap(), buy, 2, 2);
        assertThat(state.get()).isEqualTo(stateExcpected);
    }

    @Test
    @Order(135)
    public void logSession() throws Exception {
        CompletableFuture<List<SsmSessionStateLog>> sesReq = client.log(sessionName);
        List<SsmSessionStateLog> stateLog = sesReq.get();
        assertThat(stateLog.size()).isEqualTo(3);
    }

    @Test
    @Order(140)
    public void listSsm() throws Exception {
        CompletableFuture<List<String>> agentRet = client.listSsm();
        List<String> agentFormClient = agentRet.get();
        assertThat(agentFormClient).contains(ssmName);
    }

    @Test
    @Order(150)
    public void listSession() throws Exception {
        CompletableFuture<List<String>> agentRet = client.listSession();
        List<String> agentFormClient = agentRet.get();
        assertThat(agentFormClient).contains(sessionName);
    }


}

File: ssm-chaincode/ssm-chaincode-client/src/test/kotlin/ssm/chaincode/client/SsmClientItTest.kt
package ssm.chaincode.client

import com.google.common.collect.ImmutableMap
import org.assertj.core.api.Assertions
import org.assertj.core.util.Lists
import org.junit.jupiter.api.*
import org.junit.jupiter.api.MethodOrderer.OrderAnnotation
import ssm.chaincode.dsl.*
import ssm.chaincode.dsl.blockchain.Block
import ssm.chaincode.dsl.blockchain.Transaction
import ssm.sdk.sign.model.Signer
import java.util.*
import java.util.concurrent.CompletableFuture

@TestMethodOrder(OrderAnnotation::class)
class SsmClientItTest {

	companion object {
		private val uuid = UUID.randomUUID().toString()
		private const val NETWORK = "bclan-it/"
		const val ADMIN_NAME = "ssm-admin"
		val USER1_NAME = "bob" + "-" + uuid
		val USER2_NAME = "sam" + "-" + uuid
		const val USER1_FILENAME = NETWORK + "bob"
		const val USER2_FILENAME = NETWORK + "sam"

		//    private static final String NETWORK = "bc1/";
		//    private static final String ADMIN_NAME = "adrien";
		//    private static final String USER1_NAME = "chuck";
		//    private static final String USER2_NAME = "sarah";
		//    private static final String USER1_FILENAME = NETWORK+"chuck";
		//    private static final String USER2_FILENAME = NETWORK+"sarah";
		private lateinit var signerAdmin: Signer
		private lateinit var signerUser1: Signer
		private lateinit var signerUser2: Signer
		private lateinit var agentAdmin: SsmAgentBase
		private lateinit var agentUser1: SsmAgentBase
		private lateinit var agentUser2: SsmAgentBase
		private lateinit var client: SsmClient
		private lateinit var ssmName: String
		private lateinit var sessionName: String
		private lateinit var session: SsmSessionBase

		@BeforeAll
		@JvmStatic
		@Throws(Exception::class)
		fun init() {
			signerAdmin = Signer.loadFromFile(ADMIN_NAME, NETWORK + ADMIN_NAME)
			signerUser1 = Signer.loadFromFile(USER1_NAME, USER1_FILENAME)
			signerUser2 = Signer.loadFromFile(USER2_NAME, USER2_FILENAME)
			agentAdmin = loadFromFile(ADMIN_NAME, NETWORK + ADMIN_NAME)
			agentUser1 = loadFromFile(signerUser1.name, USER1_FILENAME)
			agentUser2 = loadFromFile(signerUser2.name, USER2_FILENAME)
			client = SsmClientTestBuilder.build()
			ssmName = "CarDealership-" + uuid
			val roles: Map<String, String> = ImmutableMap.of(
				signerUser1.name, "Buyer", signerUser2.name, "Seller")
			sessionName = "deal20181201-" + uuid
			session = SsmSessionBase(ssmName,
				sessionName, roles, "Used car for 100 dollars.", emptyMap())
		}

		private var privateMessage: Map<String, String>? = null
	}

	@AfterEach
	fun waitBetweenTest() {
		//Node rest api return http response before the transaction had been mined
		Thread.sleep(2000)
	}

	@Order(5)
	@Test
	fun listAdmin(): Unit {
		val agentRet = client.listAdmins()
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient).contains(ADMIN_NAME)
	}

	@Order(10)
	@Test
	fun adminUser(): Unit {
		val agentRet = client.getAdmin(ADMIN_NAME)
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient.get()).isEqualTo(agentAdmin)
	}

	@Test
	@Order(20)
	fun registerUser1() {
		val transactionEvent = client.registerUser(signerAdmin, Companion.agentUser1)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	@Order(30)
	@Test
	fun agentUser1(): Unit {
		val agentRet = client.getAgent(
			Companion.agentUser1.name)
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient.get()).isEqualTo(Companion.agentUser1)
	}

	@Test
	@Order(40)
	fun registerUser2() {
		val transactionEvent = client.registerUser(signerAdmin, Companion.agentUser2)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	@Order(50)
	@Test
	fun agentUser2(): Unit {
		val agentRet = client.getAgent(
			Companion.agentUser2.name)
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient.get()).isEqualTo(Companion.agentUser2)
	}

	@Test
	@Order(55)
	fun listAgent() {
		val agentRet = client.listAgent()
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient).contains(Companion.agentUser1.name, Companion.agentUser2.name)
	}

	@Test
	@Order(60)
	fun createSsm() {
		val sell = SsmTransitionBase(0, 1, "Seller", "Sell")
		val buy = SsmTransitionBase(1, 2, "Buyer", "Buy")
		val ssm = SsmBase(ssmName, Lists.newArrayList(sell, buy))
		val transactionEvent = client.create(signerAdmin, ssm)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	@Order(70)
	@Test
	fun ssm() {
		val ssmReq = client.getSsm(
			ssmName)
		val ssm = ssmReq.get()
		Assertions.assertThat(ssm).isPresent
		Assertions.assertThat(ssm.get().name).isEqualTo(ssmName)
	}

	@Test
	@Order(80)
	fun start() {
		val roles: Map<String, String> = ImmutableMap.of(
			Companion.agentUser1.name, "Buyer", Companion.agentUser2.name, "Seller")
		val session = SsmSessionBase(ssmName,
			sessionName, roles, "Used car for 100 dollars.", emptyMap())
		val transactionEvent = client.start(signerAdmin, session)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	@Order(90)
	@Test
	fun session() {
		val ses = client.getSession(
			sessionName)
		val sesReq = ses.get()
		Assertions.assertThat(sesReq.get().current).isEqualTo(0)
		Assertions.assertThat(sesReq.get().iteration).isEqualTo(0)
		Assertions.assertThat(sesReq.get().origin).isNull()
		Assertions.assertThat(sesReq.get().ssm).isEqualTo(ssmName)
		Assertions.assertThat(sesReq.get().roles).isEqualTo(Companion.session.roles)
		Assertions.assertThat(sesReq.get().session).isEqualTo(Companion.session.session)
		Assertions.assertThat(sesReq.get().public).isEqualTo(Companion.session.public)
	}

	@Test
	@Order(100)
	fun performSell() {
		var context = SsmContextBase(sessionName, "100 dollars 1978 Camaro", 0, emptyMap())
		context = context.addPrivateMessage("Message to signer1",
			Companion.agentUser1)
		privateMessage = context.private
		val transactionEvent = client.perform(signerUser2, "Sell", context)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	@Order(110)
	@Test
	fun sessionAfterSell(): Unit {
		val sell = SsmTransitionBase(0, 1, "Seller", "Sell")
		val sesReq = client.getSession(
			sessionName)
		val state = sesReq.get()
		val stateExpected = SsmSessionStateBase(ssmName,
			sessionName, Companion.session.roles, "100 dollars 1978 Camaro", privateMessage, sell, 1, 1)
		Assertions.assertThat(state.get()).isEqualTo(stateExpected)
		val expectedMessage = stateExpected.getPrivateMessage(signerUser1)
	}

	@Order(110)
	@Test
	fun sessionAfterSellShouldReturnEncryptMessage(): Unit {
		val (from, to, role, action) = SsmTransitionBase(0, 1, "Seller", "Sell")
		val sesReq = client.getSession(
			sessionName)
		val state = sesReq.get()
		val expectedMessage = state.get().getPrivateMessage(signerUser1)
		Assertions.assertThat(expectedMessage).isEqualTo("Message to signer1")
	}

	@Test
	@Order(120)
	fun performBuy() {
		val context = SsmContextBase(sessionName, "Deal !", 1, emptyMap())
		val transactionEvent = client.perform(signerUser1, "Buy", context)
		val trans = transactionEvent.get()
		assertThatTransactionExists(trans)
	}

	fun assertThatTransactionExists(trans: InvokeReturn) {
		Assertions.assertThat(trans).isNotNull
		Assertions.assertThat(trans.status).isEqualTo("SUCCESS")
		val transaction: Transaction = client.getTransaction(trans.transactionId).get().orElse(null)
		Assertions.assertThat(transaction).isNotNull
		Assertions.assertThat(transaction.blockId).isNotNull
		val block: Block = client.getBlock(transaction.blockId).get().orElse(null)
		Assertions.assertThat(block).isNotNull
	}

	@Order(130)
	@Test
	fun sessionAfterBuy(): Unit {
		val buy = SsmTransitionBase(1, 2, "Buyer", "Buy")
		val sesReq = client.getSession(
			sessionName)
		val state = sesReq.get()
		val stateExcpected = SsmSessionStateBase(ssmName,
			sessionName, Companion.session.roles, "Deal !", emptyMap(), buy, 2, 2)
		Assertions.assertThat(state.get()).isEqualTo(stateExcpected)
	}

	@Test
	@Order(135)
	@Throws(Exception::class)
	fun logSession() {
		val sesReq = client.log(
			sessionName)
		val stateLog = sesReq.get()
		Assertions.assertThat(stateLog.size).isEqualTo(3)
	}

	@Test
	@Order(140)
	fun listSsm() {
		val agentRet: CompletableFuture<List<String>> = client.listSsm()
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient).contains(ssmName)
	}

	@Test
	@Order(150)
	fun listSession() {
		val agentRet: CompletableFuture<List<String>> = client.listSession()
		val agentFormClient = agentRet.get()
		Assertions.assertThat(agentFormClient).contains(sessionName)
	}

}